pyccel/codegen/printing/ccode.py

```diff
@@ -1096,11 +1086,14 @@ def _print_PyccelDiv(self, expr):
 
     def _print_PyccelFloorDiv(self, expr):
         self._additional_imports.add("math")
-        if all(a.dtype is NativeInteger() for a in expr.args):
-            args = [PythonFloat(a) for a in expr.args]
-        else:
-            args = expr.args
-        code = ' / '.join(self._print(a) for a in args)
+        # the result type of the floor division is dependent on the arguments
+        # type, if all arguments are integers the result is integer otherwise
+        # the result type is float
+        need_to_cast = all(a.dtype is NativeInteger() for a in expr.args)
+        code = ' / '.join(self._print(a if a.dtype is NativeReal() else PythonFloat(a)) for a in expr.args)
+        if (need_to_cast):
+            cast_type = self.find_in_dtype_registry('int', default_precision['int'])
```

Comment:
I wonder if it would be worth using a function like `_cast_to` (see PR #714 ) to cast the result to the correct type if necessary rather than just handling native integers? I have run a quick test on my computer and FloorDiv works for numpy types too which means that you may also want one of the following:
`(float)floor(a/b)`
`(int32_t)floor(a/b)`

Reply:
No, the situation here is different, I am not casting the expression I am casting the result of the expression with an added prefix (the function name).

Reply:
I'm not sure I see what you mean here.
If I do:
```
> import numpy as np
> a = np.int32(-3)
> b = np.int32(4)
> a//b
-1
> type(a//b)
np.int32
```
Won't this be translated by pyccel to:
```
int32_t a;
int32_t b;
int32_t c;
a=-3;
b=4;
c = (int64_t)floor(-3/4); // implicit-conversion
```
?
It's a much less critical problem as the type is right but it will lead to the result being cast twice when you assign it to a variable. Or have I missed something?

Reply:
you give a great example, I need to cast to expression precision not default precision, I forgot that.
to explain more why I can't use `_cast_to` function:
I need to add cast after adding the function `(cast)func_name([expr.args])`, the _cast_to normally used to cast an expression directly, not a string, like in my case.

Reply:
Ah ok that's what you meant. I think a good improvement to the `_cast_to` function would be for it to take a string in input too or to return a statement such as `'(cast){}'` that way it could be used in a lot more places. That's why I said a function _like_ `_cast_to` ;-).

If you and @saniQQ can come up with a function that works for both of you then that is probably the most reusable version. The `_cast_to` has the advantage that it handles the test of whether the cast is necessary